starter-repos/app-template/config/base/redis/redis_manifest.yaml

```diff
@@ -462,6 +462,8 @@ apiVersion: v1
 kind: Service
 metadata:
   name: redis-master
+  # namespace has to be specified so as the kustomize can refer the metadata.namespace path.
+  namespace: default
```

Comment:
Can we avoid specifying `namespace`? Config Sync won't allow specifying multiple namespaces in a repo.

Reply:
The namespace is specified in config/dev/kustomization.yaml. That should work for kustomize. Let me know what errors you see for this part.

Reply:
kustomization.yaml has a rule that requires the existence of "redis-master" Service's metadata.namespace. @liujingfang is there a way to ignore error or use some default value?   
```yaml
vars:
- name: NAMESPACE
  objref:
    kind: Service
    name: redis-master
    apiVersion: v1
  fieldref:
    fieldpath: metadata.namespace
```

Reply:
Oh I see why you need the namespace. Since `config/envs/dev/kustomization.yaml` doesn't specify the `namespace` field. So you will see the variable not defined error. I would suggest the following:
-  add a namespace in `config/envs/dev/kustomization.yam`
- add namespace as default in `config/base/kustomization.yaml`

Reply:
I see!

Reply:
Following the user guide https://github.com/GoogleCloudPlatform/solutions-modern-cicd-anthos/blob/master/docs/index.md#testing-the-change-locally-with-skaffold

If the namespace is set to something like `default` in this config, and the user sets `kubens $USER` and then run `skaffold dev`, in which namespace will the app be deployed (`default` or `$USER`)? Would the $NAMESPACE variable be set to `default` or `$USER`?

Can we make the flow more simple instead of asking the user to open the file and edit something in it?

Reply:
For kusotmization.yaml, we expect users to manually change it such as adding resources or bases. This is different from Kptfile, which is not expected manual edits. However, since it is only for the namespace in the app-template, we can create a setter for it. Then the flow would be simplified.

Reply:
hmmm. Another option that requires zero user changes
1. namespace is excluded from ./base  
2. namespace is excluded from ./dev (so the kubens works)
3. the kustomization.yaml `configmapGenerator` and `vars` is specified in each prod-* and staging-* kustomization.yaml

@Liujingfang1 @janetkuo WDYT?

Reply:
An alternative is to provide the full config for `dev`, no overlay on ./config/base

Reply:
@janetkuo @Liujingfang1 problem solved. Found an interesting solution https://github.com/GoogleCloudPlatform/solutions-modern-cicd-anthos/pull/99/commits/1f6881c97f3673de25ee8562810d2f2945c5a849
No extra user actions are required. The dev shares the same config as staging-* and prod-* with different kustomization var setup.